agent/jvm/src/main/java/org/jolokia/jvmagent/security/KeyStoreUtil.java

```diff
@@ -137,24 +147,48 @@ private static Object createKeyPair() throws NoSuchAlgorithmException, NoSuchPro
         return keypair;
     }
 
-    private static X509Certificate getSelfCertificate(Object keypair, X500Name x500Name, Date date, long l) {
-        return (X509Certificate) ClassUtil.applyMethod(keypair, "getSelfCertificate", x500Name, date, l);
+    private static X509Certificate getSelfCertificate(Object keypair, Object x500Name, Date date, long l) {
+        final Class<?> clazz;
+        if (ClassUtil.checkForClass(X500_NAME_SUN)) {
+            clazz = ClassUtil.classForName(X500_NAME_SUN, false);
+        } else {
+            clazz = ClassUtil.classForName(X500_NAME_IBM, false);
+        }
+
+        try {
+            final Method selfCertMethod = keypair.getClass().getDeclaredMethod("getSelfCertificate", clazz, Date.class, long.class);
+            selfCertMethod.setAccessible(true);
+
+            return (X509Certificate) selfCertMethod.invoke(keypair, x500Name, date, l);
+        } catch (final NoSuchMethodException e) {
+            throw new IllegalStateException("Found no getSelfCertificate-method with the expected signature.", e);
+        } catch (final IllegalAccessException e) {
+            throw new IllegalStateException("Not allowed to access getSelfCertificate-method.", e);
+        } catch (final InvocationTargetException e) {
+            throw new IllegalStateException("The getSelfCertificate-method threw an error.", e);
+        }
     }
 
     private static PrivateKey getPrivateKey(Object keypair) {
         return (PrivateKey) ClassUtil.applyMethod(keypair, "getPrivateKey");
     }
 
     private static Class lookupKeyGenClass() {
-        Class keyGenClass = ClassUtil.classForName(KEYGEN_CLASS_JDK8);
-        if (keyGenClass == null) {
-            keyGenClass = ClassUtil.classForName(KEYGEN_CLASS_JDK7);
+        final Class keyGenClass;
+        if (ClassUtil.checkForClass(KEYGEN_CLASS_JDK8_SUN)) {
```

Comment:
Since `classForName` already return `null` when the class could not be found we could save all the calls to `checkForClass()`: Just fetch the class and if null, try next. This could be even moved out into an extra method which makes this in a loop and returns as soon as the first has been found.

Reply:
Yes, that wouldn't be a problem, at the moment I wrote the code it just didn't look very pleasing to my eye to see if this null, if that null, etc :) I can change it if you like.